File: MyCalculator/app/src/main/java/com/codewithamit/mycalculator/MainActivity.java
package com.codewithamit.mycalculator;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity<textView, result, sum> extends AppCompatActivity {

    EditText editText1;
    EditText editText2;
    TextView result;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        editText1= (EditText)findViewById(R.id.num1);
        editText2= (EditText)findViewById(R.id.num2);

        result = (TextView)findViewById(R.id.result);
    }


    public void add(View view) {

        int n1=Integer.parseInt(editText1.getText().toString());
        int n2=Integer.parseInt(editText2.getText().toString());
        int sum=n1+n2;
        result.setText(String.valueOf(sum));
    }

    public void sub(View view) {
        int n1=Integer.parseInt(editText1.getText().toString());
        int n2=Integer.parseInt(editText2.getText().toString());
        int sum=n1-n2;
        result.setText(String.valueOf(sum));
    }

    public void mul(View view) {
        int n1=Integer.parseInt(editText1.getText().toString());
        int n2=Integer.parseInt(editText2.getText().toString());
        int sum=n1*n2;
        result.setText( String.valueOf(sum));
    }

    public void div(View view) {
        int n1=Integer.parseInt(editText1.getText().toString());
        int n2=Integer.parseInt(editText2.getText().toString());
        int sum=n1/n2;
        result.setText(String.valueOf(sum));
    }
}

File: Coroutines/app/src/main/java/com/example/coroutines/MainActivity.kt
package com.example.coroutines

import android.os.AsyncTask
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private  lateinit var counterTask: CounterTask

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        btnstart.setOnClickListener {
            val n = etnumber.text.toString().toInt()
            counterTask=CounterTask()
            counterTask.execute(n)
        }

        btnstop.setOnClickListener {
            counterTask.cancel(false)
        }

    }


    inner class CounterTask : CoroutineAsyncTask<Int, Int, String>() {

        override fun doInBackground(vararg params: Int?): String {
            var count = params[0] ?: 0
            while (count >= 0) {
                if (isCancelled) {
                    return "STOPPED"
                }
                Thread.sleep(1000)
                publishProgress(count--)
            }
            return "DONE"
        }

        override fun onProgressUpdate(vararg values: Int?) {
            val progress = values[0] ?: 0
            tvready.text = progress.toString()
        }


            override fun onPostExecute(result: String?) {
            super.onProgressUpdate()
            tvready.text = result
        }

        override fun onCancelled(result: String?) {
            super.onCancelled(result)
            tvready.text = result
        }
    }
}



